xdsl/dialects/builtin.py

```diff
@@ -541,6 +541,71 @@ def verify(self, attr: Attribute) -> None:
                                | UnrankedTensorType[_VectorOrTensorElem])
 
 
+@dataclass(init=False)
+class VectorRankConstraint(AttrConstraint):
+    """
+    Constrain a vector to be of a given rank.
+    """
+
+    expected_rank: int
+    """The expected vector rank."""
+
+    def __init__(self, expected_rank: int):
+        self.expected_rank = expected_rank
+
+    def verify(self, vector: Attribute) -> None:
+        if not isinstance(vector, VectorType):
```

Comment:
This is a constraint that only allows vectors. IMO, we should find a way to not have to write this part of the verifier but instead show this information more "declaratively". So something like inherit from `BaseAttr[VectorType]`. 

Or maybe you can just leverage (ParamAttrConstraint)[https://github.com/xdslproject/xdsl/blob/main/xdsl/irdl.py#L142] to implement this?

Thoughts? @math-fehr

Reply:
I have currently no ideas on how to do that sadly, though yes we would like something like it.
I guess right now the best is to write `Annotated[VectorType[Attribute], VectorRankConstraint(...)]`, to be sure to have the right types, even though we check things twice.
However, in that `verify` itself, I don't know how to infer that information such that this is not needed.